crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1344,6 +1350,21 @@ pub struct ClassType<'db> {
     known: Option<KnownClass>,
 }
 
+#[salsa::tracked]
+impl<'db> ClassType<'db> {
+    /// Return the possible [method resolution order]s ("MRO"s) for this class.
+    ///
+    /// The MRO is the tuple of classes that can be retrieved as the `__mro__`
+    /// attribute on a class at runtime. Because the inferred type of a class base
+    /// could be a union, any class could have multiple possible MROs.
+    ///
+    /// [method resolution order]: https://docs.python.org/3/glossary.html#term-method-resolution-order
+    #[salsa::tracked(return_ref)]
+    fn mro_possibilities(self, db: &'db dyn Db) -> MroPossibilities<'db> {
```

Comment:
I expect it to be called very frequently. We need to iterate over the MRO whenever we resolve the type of an attribute or method access (explicit or implict) on a class or an instance of a class

Reply:
> Could you add an example where the resolved MRO has a cycle.

Hmmm... what are you thinking of exactly? An MRO that requires an import cycle to be resolved?

Reply:
> what are you thinking of exactly? An MRO that requires an import cycle to be resolved?

No, an actual cycle in the MRO, i.e. what you'd get if you had `class str(str): ...` in a type stub. Clearly we can't allow this and have to error on it as an invalid MRO, but it would be good to have a test showing that we do this, rather than getting stuck in an infinite loop building the MRO.

Reply:
Ahh, got it, thanks.